Math/NumberTheory/Primes/Factorisation/Montgomery.hs

```diff
@@ -152,75 +155,88 @@ stdGenFactorisation primeBound sg digits n
-                                         else repFact d' b1 b2 (count-1)
-                            (cp,cc) <- if ptest c'
-                                         then return ([(c',1)],[])
-                                         else repFact c' b1 b2 (count-1)
-                            (gp,gc) <- if ptest g
-                                         then return ([(g,2)],[])
-                                         else dbl <$> repFact g b1 b2 (count-1)
-                            return  (mergeAll [dp,cp,gp], dc ++ cc ++ gc)
+
+        ptest :: Integer -> Bool
+        ptest = maybe primeTest (\bd k -> k <= bd || primeTest k) primeBound
+
+        rndR :: Integer -> State g Integer
+        rndR k = state (prng k)
+
+        perfPw :: Integer -> (Integer, Int)
+        perfPw = maybe highestPower (largePFPower . integerSquareRoot') primeBound
+
+        fact :: Integer -> Int -> State g [(Integer, Int)]
+        fact 1 _ = return mempty
+        fact m digs = do
+          let (b1, b2, ct) = findParms digs
+          Factors pfs cfs <- repFact m b1 b2 ct
+          case cfs of
+            [] -> return pfs
+            _  -> do
+              nfs <- forM cfs $ \(k, j) ->
+                  map (second (* j)) <$> fact k (if null pfs then digs + 5 else digs)
+              return $ sconcat $ pfs :| nfs
+
+        repFact :: Integer -> Word -> Word -> Word -> State g Factors
+        repFact 1 _ _ _ = return mempty
+        repFact m b1 b2 count =
+          case perfPw m of
+            (_, 1) -> workFact m b1 b2 count
+            (b, e)
+              | ptest b   -> return $ Factors [(b, e)] []
+              | otherwise -> modifyPowers (* e) <$> workFact b b1 b2 count
+
+        workFact :: Integer -> Word -> Word -> Word -> State g Factors
+        workFact 1 _ _ _ = return mempty
+        workFact m _ _ 0 = return $ Factors [] [(m, 1)]
+        workFact m b1 b2 count = do
+          s <- rndR m
+          case s `modulo` fromInteger m of
+            InfMod{} -> error "impossible case"
+            SomeMod sm -> case montgomeryFactorisation b1 b2 sm of
+              Nothing -> workFact m b1 b2 (count - 1)
+              Just d  -> do
+                let cs = splitIntoCoprimes [(d, 1), (m `quot` d, 1)]
+                fmap mconcat $ flip mapM cs $
+                  \(x, xm) -> if ptest x
+                              then pure (Factors [(x, xm)] [])
+                              else repFact x b1 b2 (count - 1)
+
+data Factors = Factors
+  { _primeFactors     :: [(Integer, Int)]
+  , _compositeFactors :: [(Integer, Int)]
+  }
+
+-- This instance is valid only if:
+-- 1) for each argument no prime factor divides any composite factor.
+-- 2) numbers, represented by arguments, are coprime.
+instance Semigroup Factors where
+  Factors pfs1 cfs1 <> Factors pfs2 cfs2
+    = Factors (pfs1 <> pfs2) (cfs1 <> cfs2)
+
+instance Monoid Factors where
+  mempty = Factors [] []
+  mappend = (<>)
+
```

Comment:
could we make this code cleaner by considering a Generalized Version of the Factors data type thats some sort of Bifunctor/ProFunctor/SemiGroupoid/Category?

the modify Powers code is suggestive of that in my mind

Reply:
bifunctor would give bimap being the obvious one, just not sure if that improves other bits :)

Reply:
eg 
```
data Factors a b = Factors [(a,Int)] [(b,Int)] 
```
-- because we're assuming exponents x such that abs x \leq 2^31 ?

Reply:
... should/could the powers be both in Int vs Integer form for generality?

Reply:
It does not make much sense to make powers anything else then `Int`: even `2 ^ (maxBound :: Int)` is tremendously huge. So I am not sure whether it will make code clearer. 

W. r. t. to `Semigroup` / `Monoid` instances,  AFAIK there is no `DeriveMonoid` language extension to scrap this boilerplate.